### PR TITLE
Embed a default java_toolchain macro into java_tools.

### PR DESCRIPTION
**Pull request by iirina**
There are several benefits of using a macro: 
* re-using the same default values for the 2 toolchains embedded into a java_tools repository (e.g. `@java_tools//:toolchain` and `@java_tools//:toolchain_jdkX`)
* allowing users to use the embedded macro and only override parts of it (most common case is overriding `soruce_version` and `target_version`).

`@bazel_tools//tools/jdk:default_java_toolchain.bzl` will be deprecated and removed in favor of `@java_tools//:java_tools/java_tools_defs`.

This change also adds test cases for building targets in the java_tools repository. Note that ijar and singlejar don't yet build on Windows (see #8614).

### CODE REVIEW

File: src/test/shell/bazel/bazel_java_tools_test.sh [hunk 1]
```diff
@@ -173,4 +173,89 @@ function test_java_tools_has_proguard() {
...
+        echo "Skipping test." && return
+  fi
+  local java_tools_rlocation=$(rlocation io_bazel/src/java_tools_${JAVA_TOOLS_JAVA_VERSION}.zip)
+  local java_tools_zip_file_url="file://${java_tools_rlocation}"
+  if "$is_windows"; then
+        java_tools_zip_file_url="file:///${java_tools_rlocation}"
+  fi
+  cat >WORKSPACE <<EOF
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+http_archive(
```
Comment (hlopko) [inline src/test/shell/bazel/bazel_java_tools_test.sh]: Will local_repository work instead of http_archive?

File: tools/jdk/java_tools_defs.bzl [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+   "--add-exports=jdk.compiler/com.sun.tools.javac.comp=ALL-UNNAMED",
+   "--add-exports=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
+   "--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED",
+   "--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED",
+   "--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED",
+   "--add-opens=jdk.compiler/com.sun.tools.javac.file=ALL-UNNAMED",
+
+   # override the javac in the JDK.
+   "--patch-module=java.compiler=$(location :java_compiler_jar)",
+   "--patch-module=jdk.compiler=$(location :jdk_compiler_jar)",
```
Comment (hlopko) [inline tools/jdk/java_tools_defs.bzl]: I'm actually confused here, so I'll ask :) Is it true that this relative label keeps referencing `@bazel_tools//tools/jdk:java_compiler_jar` even when we call the macro from a random package `//foo/bar`? Or it assumed that the //foo/bar package has those labels around?

File: tools/jdk/java_tools_defs.bzl [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    "javac_supports_workers": 1,
+    "jvm_opts": DEFAULT_JVM_OPTS,
+    "misc": DEFAULT_JAVACOPTS,
+    "compatible_javacopts": COMPATIBLE_JAVACOPTS,
+    "singlejar": [":singlejar"],
+    "bootclasspath": ["@bazel_tools//tools/jdk:platformclasspath"],
+    "jacocorunner": ":jacoco_coverage_runner_filegroup"
+}
+
+def default_java_toolchain(name, **kwargs):
```
Comment (hlopko) [inline tools/jdk/java_tools_defs.bzl]: Alternative is to make all of these default values of attributes on java_toolchain, effectively making java_toolchain itself serve the purpose of the macro. Is that viable?

File: src/test/shell/bazel/bazel_java_tools_test.sh (hunk 1 above)
Comment (iirina) [reply src/test/shell/bazel/bazel_java_tools_test.sh]: local_repository only works if we unzip the archive first, which means additional setup and complexity in the tests. My intention was also to test the archive in the same way it is used by Bazel and Java users.

**Comment by jin** [comment]
Closing this for now, please ping/reopen if you're intending to work on this.

**Pull request closed without merging**